import math
import copy
import random

list_1 = ['a', 2, 'd', 4, 455, 67]
list_2 = ['d', 23, 'a', 45, 455, 67]

result = list(set(list_1) & set(list_2))
print(type(result))
print(f'№1----> set {result}')

result = [n for n in list_1 if n in list_2]
print(f'№1----> генератор {result}')

# 2: Дан список, заполненный произвольными числами. Получить список из элементов
# исходного, удовлетворяющих следующим условиям:
# Элемент кратен 3,
# Элемент положительный,
# Элемент не кратен 4.
#
# Примечание: Список с целыми числами создайте вручную в начале файла. Не забудьте включить туда отрицательные числа.
# 10-20 чисел в списке вполне достаточно.

my_list = [34, 2, 5, 54, -66, 6, 67, 90]
my_list1 = [n for n in my_list if n % 3 == 0 and n > 0 and n % 4 != 0]
print(f'№2----> генератор again {my_list1}')
print('////////////////////////////////////////////////////////////')
# 3. Напишите функцию которая принимает на вход список.
# Функция создает из этого списка новый список из квадратных корней чисел
# (если число положительное) и самих чисел (если число отрицательное)
# и возвращает результат (желательно применить генератор и тернарный
# оператор при необходимости).
# В результате работы функции исходный список не должен измениться.
# Например:
# old_list = [1, -3, 4]
# result = [1, -3, 2]
#
#     Примечание: Список с целыми числами создайте вручную в начале файла.
# Не забудьте включить туда отрицательные числа. 10-20 чисел в списке вполне
# достаточно.
old_list = [2, -3, 3, 0, 5, 6, -8, 2]


def do_new_list(o_l):
    new_list = o_l.copy()
    result = [round(math.sqrt(n), 2) if n >= 0 else n for n in new_list]

    # for p in new_list:
    #     if p < 0:
    #         result.append(p)

    return result


result = do_new_list(old_list)
print(f'№3-->что стало {result}')
print(f'№3--->что было {old_list}')
print('///////////////////////////////////////////////')
# 4. Написать функцию которая принимает на вход число от 1 до 100. Если число равно 13,
# функция поднимает исключительную
# ситуации ValueError иначе возвращает введенное число, возведенное в квадрат.

# Далее написать основной код программы. Пользователь вводит число. Введенное число передаем параметром
# в написанную функцию и печатаем результат, который вернула функция. Обработать возможность возникновения
# исключительной ситуации, которая поднимается внутри функции.
number = random.randint(1, 101)
print(f'№4 ---случайное число --->{number}')
number1 = int(input('Введите число: '))
print(f'№4 ---user input число --->{number1}')


def rand_excep(n):
    if n == 13:
        raise ValueError('---.№4 Введено несчастливое число ')
    else:
        result = n ** 2
        return result
# result = rand_excep(number1)
# print(f'№4----> не 13, а --{result}')
print('////////////////////////////////////////////////////////////')
try:
    result = rand_excep(number1)

except ValueError:
    print(f'-№ 4----> ОБработали ValueError  {number1} идем дальше ')
else:
    print(f'№4----> не 13 {result}')

